File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int maxFruit = 0;
        int i = 0;
        int j = 0;
        Map<Integer, Integer> fruitCount = new HashMap<>();

        while (j < fruits.length) {
            fruitCount.put(fruits[j], fruitCount.getOrDefault(fruits[j], 0) + 1);

            while (fruitCount.size() > 2) {
                fruitCount.put(fruits[i], fruitCount.get(fruits[i]) - 1);
                if (fruitCount.get(fruits[i]) == 0) {
                    fruitCount.remove(fruits[i]);
                }
                i++;
            }

            maxFruit = Math.max(maxFruit, j - i + 1);
            j++;
        }

        return maxFruit;
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
import collections
from typing import List

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        count = collections.defaultdict(int)
        left, currTotal, maxi = 0, 0, 0
        
        for right in range(len(fruits)):
            count[fruits[right]] += 1
            currTotal += 1
            
            while len(count) > 2:
                count[fruits[left]] -= 1
                currTotal -= 1
                if count[fruits[left]] == 0:
                    count.pop(fruits[left])
                left += 1
            
            maxi = max(currTotal, maxi)
        
        return maxi
